include/gmsh.tcc

```diff
@@ -4,137 +4,151 @@
+      break;
+    }
+  }
+  if (!read_status) {
+    std::cerr << "Cannot find keyword: " << keyword << '\n';
+  }
 }
 
 //! Read vertex id and coordinates in GMSH
 //! \tparam Tdim Dimension
 //! \tparam Tvertices Number of vertices in element
 //! \param[in] filename Input mesh filename
 template <unsigned Tdim, unsigned Tvertices>
-void GMSH<Tdim, Tvertices>::read_vertices(const std::string& filename) {
+void GMSH<Tdim, Tvertices>::read_vertices(std::ifstream& file) {
+
+  //! Find the line of interest
+  read_keyword(file, "$Nodes");
 
-  //! Number of vertices
-  double nvertices = std::numeric_limits<double>::max();
-  const unsigned toplines = 4;
+  std::string line;
+  std::getline(file, line);
+  std::istringstream istream(line);
+
+  //! Read number of vertices
+  unsigned nvertices = std::numeric_limits<unsigned>::max();
+  istream >> nvertices;
+  getline(istream, line);
 
   //! Vertices id
   unsigned vertid = 0;
 
-  std::fstream infile;
-  infile.open(filename, std::ios::in);
-
-  //! Check if input file is good
-  if (infile.good()) {
-    std::cout << "Vertices file found" << '\n';
+  //! Read vertex coordinates & id
+  for (unsigned i = 0; i < nvertices; ++i) {
+    std::getline(file, line);
+    std::istringstream istream(line);
 
-    std::string line;
+    if (line.find('#') == std::string::npos && line != "") {
+      //! Coordinates of vertex
+      std::array<double, Tdim> vertex;
 
-    //! Ignore first 4 lines
-    for (unsigned i = 0; i < toplines; ++i) {
-      std::getline(infile, line);
-    }
-    //! Read number of vertices
-    infile >> nvertices;
-    getline(infile, line);
-
-    //! Read vertex coordinates & id
-    for (unsigned i = 0; i < nvertices; ++i) {
-      std::getline(infile, line);
-      std::istringstream istream(line);
-
-      if (line.find('#') == std::string::npos && line != "") {
-        //! Coordinates of vertex
-        std::array<double, Tdim> vertex;
-
-        istream >> vertid;
-        istream >> vertex.at(0) >> vertex.at(1) >> vertex.at(2);
-        vertices_.insert(std::make_pair(vertid, vertex));
-      }
+      istream >> vertid;
+      istream >> vertex.at(0) >> vertex.at(1) >> vertex.at(2);
+      this->vertices_.insert(std::make_pair(vertid, vertex));
     }
-    infile.close();
   }
-  std::cout << "Number of Vertices: " << vertices_.size() << '\n';
+
+  this->nvertices_ = vertices_.size();
+
+  //! Check that the number of vertices are correct
+  if (nvertices_ != nvertices) {
+    std::cout << "Error in the code, number of vertices do not match.\n";
+  }
+
+  std::cout << "Number of Vertices: " << nvertices_ << '\n';
 }
 
 //! Read GMSH elements
 //! \tparam Tdim Dimension
 //! \tparam Tvertices Number of vertices in element
 //! \param[in] filename Input mesh filename
 template <unsigned Tdim, unsigned Tvertices>
-void GMSH<Tdim, Tvertices>::read_elements(const std::string& filename) {
+void GMSH<Tdim, Tvertices>::read_elements(std::ifstream& file) {
```

Comment:
Function definition doesn't match the comments.

Reply:
can you be more specific? is it because there is no directory for the filename?